File: Data Structures & Algorithms/target-sum/submission-1.cpp
class Solution {
public:
    int req(int amount, int pos,vector<int>&coins,vector<vector<int>>&dp){
        if(amount==0 and pos == coins.size()){
            return 1;
        }

        if(pos==coins.size()){
            return 0;
        }
        if(dp[amount+20000][pos]!=-1){
            return dp[amount+20000][pos];
        }
        return dp[amount+20000][pos]=req(amount-coins[pos],pos+1,coins,dp)+req(amount+coins[pos],pos+1,coins,dp);
    }
    int findTargetSumWays(vector<int>& nums, int target) {
        vector<vector<int>> dp(40000,vector<int>(nums.size(),-1));
        // dp[0][0] = 0;
        return req(target,0,nums,dp);
    }
};

File: Data Structures & Algorithms/find-duplicate-integer/submission-1.py
class Solution:
    def findDuplicate(self, nums: List[int]) -> int:
        
        ans = -1
        i = 0
        for i in range(len(nums)):
            if nums[i]>=0:
                if nums[nums[i]]<0:
                    return nums[i]
                else:
                    nums[nums[i]] = -nums[nums[i]]
            else:
                pos = -nums[i]
                if nums[pos]<0:
                    return pos
                else:
                    nums[pos] = -nums[pos]
        